src/types.rs

```diff
@@ -104,6 +120,21 @@ macro_rules! key_methods {
                 // access
                 &*(bytes.as_ptr() as *const Self)
             }
+
+            /// Copy-free reference cast
+            pub fn as_ref_checked(bytes: &[u8]) -> Option<&Self> {
+                let inner: &[u8; Self::LEN] = bytes.try_into().ok()?;
+
+                // The interpreter will frequently make references to keys and values using
+                // logically checked slices.
+                //
+                // This function will save unnecessary copy to owned slices for the interpreter
+                // access
+                //
+                // The size is checked above and Self is `#[repr(transparent)]`
+                // so the layout is guaranteed to be correct.
+                Some(unsafe { &*(inner.as_ptr() as *const Self) })
```

Comment:
We are using `unsafe` inside a block that is not shielded by `unsafe_rust`. Does that make sense? This looks like a redundant function with just the len check but same critical constraints of its unchecked counterpart

Reply:
No because the only unsafe thing the caller could do before was call this with the wrong length. There's nothing the caller can do wrong to create unsafe behavior so it's ok to put within a safe api.

Reply:
So we came to the conclusion to make this the new unsafe function and put it behind the `optimized` feature as it still has unsafe in it. It does introduce a bounds check but I feel that is really worth it here as getting it wrong can segfault or worse. 
We could create a safe api that uses unit structs to prove the check was done somewhere else and allow the bounds check to be skipped but that feels like an optimization we should do if we want to squeeze the extra performance out of this function.
When the `optimized` feature is off the `from_slice` can be used to create an owned value and take a reference to that. 
@vlopes11 Are you ok with this?

Reply:
That sounds good as long as we preserve the current implementation of the VM inside of a feature of choice. The reason for that is this bounds check is redundant to the specs. If all the arithmetic overflows and instructions requirements passes, then the out of bounds error is unreachable.

As long as this change is incremental and won't override the implementation of the slice to type cast, we are good to go.